soroban-env-host/src/native_contract/token/balance.rs

```diff
@@ -6,74 +6,150 @@ use crate::native_contract::token::storage_types::DataKey;
 use core::cmp::Ordering;
 use soroban_env_common::xdr::{
-    AccountEntryExt, AccountEntryExtensionV1Ext, AccountId, LedgerEntryData, TrustLineAsset,
-    TrustLineEntryExt, TrustLineFlags,
+    AccountEntryExt, AccountEntryExtensionV1Ext, AccountFlags, AccountId, LedgerEntryData,
+    TrustLineAsset, TrustLineEntryExt, TrustLineFlags,
 };
 use soroban_env_common::{CheckedEnv, TryIntoVal};
 
 use super::error::ContractError;
+use super::storage_types::BalanceValue;
 
 // Metering: *mostly* covered by components. Not sure about `try_into_val`.
 pub fn read_balance(e: &Host, id: Identifier) -> Result<BigInt, HostError> {
     let key = DataKey::Balance(id);
     if let Ok(balance) = e.get_contract_data(key.try_into_val(e)?) {
-        Ok(balance.try_into_val(e)?)
+        let balance: BalanceValue = balance.try_into_val(e)?;
+        Ok(balance.amount)
     } else {
         Ok(BigInt::from_u64(e, 0)?)
     }
 }
 
+fn is_issuer_auth_required(e: &Host, issuer_id: AccountId) -> Result<bool, HostError> {
+    let lk = e.to_account_key(issuer_id.clone());
+    e.with_mut_storage(|storage| {
+        //TODO: Should we add a try_get method to storage?
+        if storage.has(&lk)? {
+            let le = storage.get(&lk)?;
+            let auth_required = match le.data {
+                LedgerEntryData::Account(ae) => {
+                    (ae.flags & (AccountFlags::RequiredFlag as u32)) != 0
+                }
+                _ => false, //TODO: This shouldn't happen... should we error instead?
+            };
+
+            Ok(auth_required)
+        } else {
+            //issuer is missing, so authorization is not required because anyone
+            //can re-create the issuer without issuer flags
+            Ok(false)
+        }
+    })
+}
+
+fn is_asset_auth_required(e: &Host) -> Result<bool, HostError> {
+    match read_metadata(e)? {
+        Metadata::Token(_) => Ok(false),
+        Metadata::Native => Ok(false),
+        Metadata::AlphaNum4(asset) => is_issuer_auth_required(e, asset.issuer),
+        Metadata::AlphaNum12(asset) => is_issuer_auth_required(e, asset.issuer),
+    }
+}
+
 // Metering: *mostly* covered by components. Not sure about `try_into_val`.
 fn write_balance(e: &Host, id: Identifier, amount: BigInt) -> Result<(), HostError> {
-    let key = DataKey::Balance(id);
-    e.put_contract_data(key.try_into_val(e)?, amount.try_into_val(e)?)?;
+    let raw_key = DataKey::Balance(id).try_into_val(e)?;
+
+    if let Ok(raw_balance) = e.get_contract_data(raw_key) {
+        let mut balance: BalanceValue = raw_balance.try_into_val(e)?;
+        balance.amount = amount;
+        e.put_contract_data(raw_key, balance.try_into_val(e)?)?;
+    } else {
+        // Balance does not exist, so now we need to enforce auth_required IF this is a wrapped token
+        // with an auth_required issuer.
+        if is_asset_auth_required(e)? {
+            return Err(e.err_status_msg(
+                ContractError::IssuerAuthRequiredError,
```

Comment:
Do we still need to return the error in case if the balance comes to an account with an authorized trustline. I.e., should the user be able to `import` the auth required asset for which they already received authorization?

Reply:
This is a great question. I'm currently assuming that auth_required issuers might want to opt trustlines into soroban even if they've already been authorized on classic. Ultimately, this should be determined by requirements. If no partner cares about re-authorizing balances that have authorized trustlines, then we should authorize the balance by default. Would be good to get your thoughts @JakeUrban and @tomerweller.

Reply:
Just make it clear: even if we auto-authorize the token on soroban, it still will still only be transferrable only between the authorized identifiers on Soroban (either auto-authorized or explicitly authorized on Soroban), right? If so, then this seems pretty much equivalent to classic in terms of the scope of authorization.

Reply:
Yeah that's right.

Reply:
I agree this is a good question. I can get feedback on this from our auth-required partners, but I suspect that auto-authorization of balances is acceptable, and maybe even preferable, in the case where the account's trustline is authorized.

Personally, I think asset issuers are going to be confused / frustrated with the fact that their users have two distinct balances of their asset, so anything we can do to make it feel like a single-balance experience will likely be well received.

Question: I see that the issuer must authorize identifiers. Identifiers include contracts right? So auth required issuers would have to explicitly authorize usage of specific contracts right?

Reply:
Makes sense. The single-balance discussions we've been having affect this, so I'll leave it as is for now. We can add auto-authorization if we decide to keep split balances.

> Question: I see that the issuer must authorize identifiers. Identifiers include contracts right? So auth required issuers would have to explicitly authorize usage of specific contracts right?

Correct.